internal/overlord/checkstate/manager.go

```diff
@@ -0,0 +1,216 @@
+// stopping the previous checks and starting the new ones as required.
+func (m *CheckManager) Configure(p *plan.Plan) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
+		len(m.checks), len(p.Checks))
+
+	// First stop existing checks.
+	for _, check := range m.checks {
+		check.stop()
+	}
+
+	// Then configure and start new checks.
+	checks := make(map[string]*checkData, len(p.Checks))
+	for name, config := range p.Checks {
+		check := &checkData{
+			config:  config,
+			checker: newChecker(config),
+			done:    make(chan struct{}),
+			action:  m.callFailureHandlers,
+		}
+		checks[name] = check
+		go check.loop()
+	}
+	m.checks = checks
+}
+
+func (m *CheckManager) callFailureHandlers(name string) {
+	for _, f := range m.failureHandlers {
+		f(name)
+	}
+}
+
+// Checks returns the list of currently-configured checks and their status,
+// ordered by name.
+//
+// If level is not UnsetLevel, the list of checks is filtered to only include
+// checks with the given level. If names is non-empty, the list of checks is
+// filtered to only include the named checks.
+func (m *CheckManager) Checks(level plan.CheckLevel, names []string) ([]*CheckInfo, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	var infos []*CheckInfo
+	for _, check := range m.checks {
+		if level != plan.UnsetLevel && level != check.config.Level {
+			continue
+		}
+		if len(names) > 0 && !containsString(names, check.config.Name) {
+			continue
+		}
+		infos = append(infos, check.info())
+	}
+
+	sort.Slice(infos, func(i, j int) bool {
+		return infos[i].Name < infos[j].Name
+	})
+	return infos, nil
+}
+
+func containsString(slice []string, value string) bool {
+	for _, v := range slice {
+		if v == value {
+			return true
+		}
+	}
+	return false
+}
+
+// CheckInfo provides status information about a single check.
+type CheckInfo struct {
+	Name         string
+	Healthy      bool
+	Failures     int
+	LastError    string
+	ErrorDetails string
+}
+
+// checkData holds state for an active health check.
+type checkData struct {
+	config  *plan.Check
+	checker checker
+	done    chan struct{}
+
+	mutex     sync.Mutex
+	failures  int
+	action    FailureFunc
+	actionRan bool
+	lastErr   error
+	cancel    context.CancelFunc
+}
+
+type checker interface {
+	check(ctx context.Context) error
+}
+
+func (c *checkData) loop() {
+	logger.Debugf("Check %q starting with period %v", c.config.Name, c.config.Period.Value)
+	defer logger.Debugf("Check %q stopped", c.config.Name)
+
+	ticker := time.NewTicker(c.config.Period.Value)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.check()
+		case <-c.done:
+			return
+		}
+	}
+}
+
+func (c *checkData) check() {
+	// Set up context: a timeout or call to c.stop() will cancel the check.
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.Timeout.Value)
+	defer cancel()
+	c.mutex.Lock()
+	c.cancel = cancel
```

Comment:
checkData should have its own context.Context

`c.ctx, c.cancel = context.WithCancel(context.Background())`

then check()'s ctx is single purpose
```
cox, cancel := context.WithTimeout(c.ctx, c.config.Timeout.Value)
defer cancel()
```

Reply:
Ah, this is a good call -- simplifies things a bit and avoids some extra locking.